plugins/home/src/api/VisitsStorageApi.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { IdentityApi, StorageApi } from '@backstage/core-plugin-api';
+import {
+  Visit,
+  VisitsApi,
+  VisitsApiQueryParams,
+  VisitsApiSaveParams,
+} from './VisitsApi';
+
+/** @public */
+export type VisitsStorageApiOptions = {
+  limit?: number;
+  storageApi: StorageApi;
+  identityApi: IdentityApi;
+};
+
+type ArrayElement<A> = A extends readonly (infer T)[] ? T : never;
+
+/**
+ * @public
+ * This is an implementation of VisitsApi that relies on a StorageApi.
+ * Beware that filtering and ordering are done in memory therefore it is
+ * prudent to keep limit to a reasonable size.
+ */
+export class VisitsStorageApi implements VisitsApi {
```

Comment:
```suggestion
export class VisitsStorageApiFactory implements VisitsApi {
```

Reply:
Suggestion (non-blocking): Appending a factory suffix sounds more clear to me.

Reply:
I've used the term factory before in a loosly sense. I've decided to remove it because it did not match the way it is used in literarture as far as I could see (cf. [abstract factory](https://wiki.c2.com/?AbstractFactory) or [factory method pattern](https://wiki.c2.com/?FactoryMethodPattern)).
In this case I have a class `VisitsStorageApi` with an static method that returns an object of the same type. That wouldn't be best described as a factory.
Correct me if I'm wrong. I understand your intention here is to merge with you previous point. Removing VisitsWebStorageApi and creating something like a VisitsApiFactory. So I could have methods for both. An example bellow:
```typescript
class VisitsStorageApi implements VisitsApi {
// (...)
}

export class VisitsApiFactory {
// (...)
static createFromStorageApi(options) { VisitsStorageApi.create(...) }
static createWithWebStorage(options) { VisitsStorageApi.create(...) }
// (...)
}
```
Whitch could be a viable option.

Reply:
Almost, my suggestion is that `WebVisitsStorage` be removed and `VisitsStorageApi` can take an implementation of `storageApi` as a dependency, and that implementation can be a `WebStorage`. With this, we can get rid of the "retrive all" and "persist all" methods because if I understand correctly, if we only depend on `storageApi`, we will always recover and persist the data in the same way.
Does it make sense?

Reply:
Resolving this conversation since the suggestion is already in code.